### DISCUSSION TITLE
Transitioning back the a previous state

### DISCUSSION THREAD

**Discussion by karlmax-elcabron-dev** [imjp94/gd-YAFSM#13] [Q&A] [answered]
Hello, first of all I want to say: great work, I really like your plugin! I love to integrate my paper sketches in the editor with your UI and I like the triggers/parameters mechanism (my previous state machine was little more than a stack that I controlled from my game code)

My gamedev experience is basic and all self taught, so maybe my question could sound silly to you. Let's take the example of an Ant's fsm:

![image](https://user-images.githubusercontent.com/81615431/112962099-4c5f8a00-9146-11eb-9fb2-c86787c480db.png)

The "Flee" state can be reached transiting both from "FindFood" and "GoHome" states, when some danger is near the ant. But when I set trigger "danger_is_far" to go back to the previous state, my ant goes anyhow back to the "FindFood" entry state.

Should I manage the stack manually in this case, or is it automated and I simply am missing something?

Thank you for your work and your attention

**Comment by imjp94** [comment]
Thank you!

This issue is not silly at all, in fact, it address the current limitation of this plugin - no way to sort transition.
Currently(0.4.1), transitions are saved in an unordered manner, so you can't manually control the priority of transition when `StateMachinePlayer` playing.
And that's why triggering "danger_is_far" always transit back to "FindFood"(regardless of whether the last state is "GoHome" or not)

Current workaround would be using different parameter name, for example, "safe_to_go_home" or "safe_to_find_food" to avoid this racing issue.

I always thought priority in transition doesn't matter that much, but now, I would consider adding it on next update.
You can track the issue from #14

**Comment by imjp94** [comment, marked as answer]
Oh wait... Sorting transition doesn't actually solve this racing issue, it will only make it biased to the transition with higher priority...

This issue is more of a flaw in the design, as "Flee" is responsible of making decision whether to "GoHome" or "FindFood", so using different parameter name, for example, "safe_to_go_home" or "safe_to_find_food" should be the cleanest way to achieve what you are trying to do.

**Reply by karlmax-elcabron-dev** [reply]
Ok, thank you. So in this case if I differentiate the two triggers, I need to check the smp.get_previous() to be able to decide which trigger to send to go back to the previous state.

In my own previous, yet very basic and very spaghetti implementation of a fsm I did this:

**GoHome** or **FindFood** --> *danger_is_near* --> push state **Flee** --> *danger_is_far* --> pop state **Flee**

to go back to the state I was before fleeing, so I maintained that design also when using your plugin. 

I repeat, I am a tinkerer and completely self taught, maybe I should study some design patterns some day :)

Thank you for your answers and have a nice day

**Reply by imjp94** [reply]
> So in this case if I differentiate the two triggers, I need to check the smp.get_previous() to be able to decide which trigger to send to go back to the previous state.

Yes

> GoHome or FindFood --> danger_is_near --> push state Flee --> danger_is_far --> pop state Flee

You are right about it, that's how I do it before I created this plugin too, in fact, `StateMachinePlayer` is backed by [Stack](https://github.com/imjp94/gd-YAFSM/blob/master/addons/imjp94.yafsm/src/StackPlayer.gd) as well.

But unlike a "vanilla" stack-based fsm, gd-YAFSM's `StateMachine` are changing state with the helps of transitions, so it will loop through all the connecting transitions and see where it is going next.

In short, 
For "vanilla" stack-based fsm, you tell it where to go directly through `pop` or `push`. 
But for case of gd-YAFSM's `StateMachine`, you input parameters to `StateMachinePlayer` and let the "flowchart" decides where to go, sort of like some kind of automation.
